File: testing/acceptance/temperature-spa/shared/runner/templates.ts
// testing/acceptance/temperature-spa/shared/runner/templates.ts
//
// gm-root.27.2 — bead-template handler registry.
//
// Each handler is a `(ctx, bead) => Promise<void>` that performs the
// mechanical work the matching bead frontmatter declares. The mock
// runner picks the handler by name (frontmatter `template:` field)
// and invokes it.
//
// The handlers are deliberately specific to the temperature-spa
// target. There's no generic "write-component" — write-component
// dispatches to a per-file content registry keyed by the bead's
// `files:` frontmatter. The registry lives below in CONTENT_REGISTRY.
//
// References:
//   - D16 docs/design/acceptance-temperature-spa.md §4 (template list)
//   - D17 docs/design/acceptance-temperature-spa.md §5 (per-bead frontmatter)

import { spawnSync, type SpawnSyncReturns } from 'node:child_process';
import { mkdirSync, writeFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import type { Frontmatter } from './frontmatter';

export type TemplateContext = {
  /** Absolute path to the target project's root. */
  projectDir: string;
  /** The bead being worked. */
  beadID: string;
  /** The bead's labels — used to disambiguate per-milestone variants. */
  labels: string[];
  /**
   * Logger callback so the mock runner can stream template progress
   * into the test report. Optional; defaults to no-op.
   */
  log?: (msg: string) => void;
};

export type TemplateHandler = (
  ctx: TemplateContext,
  fm: Frontmatter
) => Promise<void>;

/**
 * Look up a handler by name. Returns undefined if no template
 * matches; the caller (mock runner) escalates with template_unknown.
 */
export function getTemplate(name: string | undefined): TemplateHandler | undefined {
  if (!name) return undefined;
  return TEMPLATES[name];
}

const TEMPLATES: Record<string, TemplateHandler> = {
  'init-repo': initRepo,
  'npm-install': npmInstall,
  'write-component': writeComponent,
  'write-test': writeComponent, // same path: write file from registry
  build: build,
  serve: serveTemplate,
  'error-then-recover': errorThenRecover,
  noop: noop,
};

// ─────────────────────────────────────────────────────────────────────
// Templates
// ─────────────────────────────────────────────────────────────────────

async function initRepo(ctx: TemplateContext, fm: Frontmatter): Promise<void> {
  ctx.log?.(`init-repo: ${ctx.beadID}`);
  // git init.
  runSync('git', ['init', '--initial-branch=main'], ctx.projectDir);
  // Write each declared file from the content registry.
  await writeRegistryFiles(ctx, fm);
}

async function npmInstall(ctx: TemplateContext): Promise<void> {
  ctx.log?.(`npm-install: ${ctx.beadID}`);
  // Prefer offline cache (D16 §4.3). If --offline fails (cache miss),
  // fall through to a regular install — keeps the test resilient when
  // the cache hasn't been pre-warmed yet.
  const offline = spawnSync('npm', ['install', '--offline', '--silent'], {
    cwd: ctx.projectDir,
    stdio: 'pipe',
  });
  if (offline.status !== 0) {
    ctx.log?.('npm-install: offline cache miss, falling back to online');
    runSync('npm', ['install', '--silent'], ctx.projectDir);
  }
}

async function writeComponent(ctx: TemplateContext, fm: Frontmatter): Promise<void> {
  ctx.log?.(`write-component: ${ctx.beadID} files=${(fm.files ?? []).join(',')}`);
  await writeRegistryFiles(ctx, fm);
}

async function build(ctx: TemplateContext): Promise<void> {
  ctx.log?.(`build: ${ctx.beadID}`);
  runSync('npm', ['run', 'build'], ctx.projectDir);
}

async function serveTemplate(ctx: TemplateContext): Promise<void> {
  // serve-template is a marker — actual preview-server lifecycle is
  // owned by the M2/M3 step beads (gm-root.27.9 / .10), which spawn
  // and tear down vite preview around their oracle assertions. The
  // template's job is just to verify dist/ is buildable.
  ctx.log?.(`serve: ${ctx.beadID} (no-op; preview owned by step bead)`);
}

let errorThenRecoverCount = 0;
async function errorThenRecover(ctx: TemplateContext): Promise<void> {
  ctx.log?.(`error-then-recover: attempt ${errorThenRecoverCount + 1}`);
  errorThenRecoverCount += 1;
  if (errorThenRecoverCount === 1) {
    throw new Error(
      'error-then-recover: deliberate first-attempt failure (gm-root.27.2)'
    );
  }
}

async function noop(ctx: TemplateContext, fm: Frontmatter): Promise<void> {
  // Drop in any files declared in the frontmatter. Useful for "just
  // copy these config files" beads (e.g., M1.2 which writes
  // tsconfig.json + index.html + src/main.tsx).
  ctx.log?.(`noop: ${ctx.beadID} files=${(fm.files ?? []).join(',')}`);
  await writeRegistryFiles(ctx, fm);
}

// ─────────────────────────────────────────────────────────────────────
// Helpers
// ─────────────────────────────────────────────────────────────────────

async function writeRegistryFiles(
  ctx: TemplateContext,
  fm: Frontmatter
): Promise<void> {
  const files = fm.files ?? [];
  if (files.length === 0) return;
  for (const relPath of files) {
    if (relPath.endsWith('/')) {
      // Directory marker (e.g., 'node_modules/'); not a file we
      // synthesize — skip silently. Templates like npm-install
      // produce these via real commands.
      continue;
    }
    const content = resolveContent(relPath, ctx.labels, ctx.beadID);
    if (content === undefined) {
      throw new Error(
        `MockAgentRunner: no registry entry for file '${relPath}' (bead ${ctx.beadID})`
      );
    }
    const abs = join(ctx.projectDir, relPath);
    mkdirSync(dirname(abs), { recursive: true });
    writeFileSync(abs, content, 'utf8');
  }
}

function resolveContent(
  relPath: string,
  labels: string[],
  beadID: string
): string | undefined {
  // App.tsx has two versions: the M2 "Hello world" and the M3
  // "render TemperatureTable". Disambiguate by milestone label.
  if (relPath === 'src/App.tsx') {
    if (labels.includes('milestone:m3')) return CONTENT_REGISTRY['src/App.tsx#m3'];
    return CONTENT_REGISTRY['src/App.tsx#m2'];
  }
  return CONTENT_REGISTRY[relPath];
}

/**
 * Static content for every file the target SPA contains. Authoring
 * is hand-tuned to the D17 contract (testids + frontmatter) and the
 * D18 oracle (numeric correctness, formatted-string °F).
 *
 * If the M3 oracle fails on a row mismatch, look here first — this
 * is the source of truth the mock writes.
 */
const CONTENT_REGISTRY: Record<string, string> = {
  'package.json': JSON.stringify(
    {
      name: 'temperature-spa',
      private: true,
      version: '0.0.0',
      type: 'module',
      scripts: {
        dev: 'vite',
        build: 'tsc --noEmit && vite build',
        preview: 'vite preview',
        test: 'vitest run',
      },
      dependencies: {
        react: '^18.3.1',
        'react-dom': '^18.3.1',
      },
      devDependencies: {
        '@testing-library/react': '^16.1.0',
        '@types/react': '^18.3.12',
        '@types/react-dom': '^18.3.1',
        '@vitejs/plugin-react': '^4.3.4',
        jsdom: '^25.0.1',
        typescript: '^5.6.3',
        vite: '^5.4.11',
        vitest: '^2.1.8',
      },
    },
    null,
    2
  ) + '\n',
  'vite.config.ts':
    `import { defineConfig } from 'vite';
import react from '@vitejs/plugin-react';

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [react()],
  test: {
    environment: 'jsdom',
    globals: true,
  },
});
`,
  'tsconfig.json': JSON.stringify(
    {
      compilerOptions: {
        target: 'ES2022',
        lib: ['ES2022', 'DOM', 'DOM.Iterable'],
        module: 'ESNext',
        moduleResolution: 'Bundler',
        jsx: 'react-jsx',
        strict: true,
        esModuleInterop: true,
        resolveJsonModule: true,
        skipLibCheck: true,
        allowImportingTsExtensions: false,
        isolatedModules: true,
        noEmit: true,
      },
      include: ['src'],
    },
    null,
    2
  ) + '\n',
  'index.html':
    `<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Temperature SPA</title>
  </head>
  <body>
    <div id="root"></div>
    <script type="module" src="/src/main.tsx"></script>
  </body>
</html>
`,
  'src/main.tsx':
    `import React from 'react';
import ReactDOM from 'react-dom/client';
import App from './App';

const root = document.getElementById('root');
if (!root) throw new Error('root element not found');
ReactDOM.createRoot(root).render(<React.StrictMode><App /></React.StrictMode>);
`,
  'src/App.tsx#m2':
    `export default function App() {
  return <div data-testid="app-root">Hello world</div>;
}
`,
  'src/App.tsx#m3':
    `import TemperatureTable from './TemperatureTable';

export default function App() {
  return (
    <div data-testid="app-root">
      <TemperatureTable />
    </div>
  );
}
`,
  'src/App.test.tsx':
    `import { describe, it, expect } from 'vitest';
import { render, screen } from '@testing-library/react';
import App from './App';

describe('App', () => {
  it('renders Hello world inside app-root', () => {
    render(<App />);
    const root = screen.getByTestId('app-root');
    expect(root.textContent).toBe('Hello world');
  });
});
`,
  'src/temperatureRows.ts':
    `// 16 rows for °C ∈ {0, 20, 40, ..., 300}. °F = °C × 9/5 + 32, one decimal.
export interface TemperatureRow {
  celsius: number;
  fahrenheit: string;
}

export function temperatureRows(): TemperatureRow[] {
  const rows: TemperatureRow[] = [];
  for (let c = 0; c <= 300; c += 20) {
    rows.push({
      celsius: c,
      fahrenheit: ((c * 9) / 5 + 32).toFixed(1),
    });
  }
  return rows;
}
`,
  'src/TemperatureTable.tsx':
    `import { temperatureRows } from './temperatureRows';

export default function TemperatureTable() {
  const rows = temperatureRows();
  return (
    <table data-testid="temperature-table">
      <thead>
        <tr>
          <th>Celsius</th>
          <th>Fahrenheit</th>
        </tr>
      </thead>
      <tbody>
        {rows.map((r) => (
          <tr key={r.celsius} data-testid={\`row-\${r.celsius}\`}>
            <td>{r.celsius}</td>
            <td>{r.fahrenheit}</td>
          </tr>
        ))}
      </tbody>
    </table>
  );
}
`,
  'src/TemperatureTable.test.tsx':
    `import { describe, it, expect } from 'vitest';
import { render, screen } from '@testing-library/react';
import TemperatureTable from './TemperatureTable';

describe('TemperatureTable', () => {
  it('renders exactly 16 rows', () => {
    render(<TemperatureTable />);
    const rows = screen.getAllByTestId(/^row-/);
    expect(rows).toHaveLength(16);
  });

  it('row-0 shows 32.0', () => {
    render(<TemperatureTable />);
    const row = screen.getByTestId('row-0');
    expect(row.textContent).toContain('0');
    expect(row.textContent).toContain('32.0');
  });

  it('row-100 shows 212.0', () => {
    render(<TemperatureTable />);
    const row = screen.getByTestId('row-100');
    expect(row.textContent).toContain('100');
    expect(row.textContent).toContain('212.0');
  });

  it('row-300 shows 572.0', () => {
    render(<TemperatureTable />);
    const row = screen.getByTestId('row-300');
    expect(row.textContent).toContain('300');
    expect(row.textContent).toContain('572.0');
  });
});
`,
};

function runSync(cmd: string, args: string[], cwd: string): void {
  const res: SpawnSyncReturns<Buffer> = spawnSync(cmd, args, {
    cwd,
    stdio: 'pipe',
  });
  if (res.status !== 0) {
    const err = res.stderr?.toString() ?? '';
    throw new Error(
      `MockAgentRunner: \`${cmd} ${args.join(' ')}\` exited ${res.status} in ${cwd}\n${err}`
    );
  }
}

File: internal/adapter/mock/templates.go
// templates.go (gm-root.28.5) — port of
// testing/acceptance/temperature-spa/shared/runner/templates.ts.
//
// Eight handlers and a per-file content registry hand-tuned to the
// D17 (gm-xw8a) JSONL pack and the D18 (gm-bvlm) numeric oracle.
//
// Each handler receives a TemplateContext (projectDir, beadID, labels,
// log) and a Frontmatter (template, testid, files, extras). Handlers
// perform the mechanical work the matching bead frontmatter declares
// — writing files, shelling out to npm/vite, etc.

package mock

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"sync"
)

// TemplateContext is threaded through every handler.
type TemplateContext struct {
	ProjectDir string
	BeadID     string
	Labels     []string
	Log        func(string)
}

func (c TemplateContext) log(msg string) {
	if c.Log != nil {
		c.Log(msg)
	}
}

// TemplateHandler is the signature every template implements.
type TemplateHandler func(ctx TemplateContext, fm Frontmatter) error

// GetTemplate returns a handler by name. Returns nil if unknown.
// Mirrors the TS getTemplate.
func GetTemplate(name string) TemplateHandler {
	switch name {
	case "init-repo":
		return initRepo
	case "npm-install":
		return npmInstall
	case "write-component", "write-test":
		return writeComponent
	case "build":
		return buildTpl
	case "serve":
		return serveTpl
	case "error-then-recover":
		return errorThenRecover
	case "noop":
		return noop
	default:
		return nil
	}
}

// ─── Handlers ──────────────────────────────────────────────────────

func initRepo(ctx TemplateContext, fm Frontmatter) error {
	ctx.log(fmt.Sprintf("init-repo: %s", ctx.BeadID))
	if err := runCmd(ctx.ProjectDir, "git", "init", "--initial-branch=main"); err != nil {
		return err
	}
	return writeRegistryFiles(ctx, fm)
}

func npmInstall(ctx TemplateContext, _ Frontmatter) error {
	ctx.log(fmt.Sprintf("npm-install: %s", ctx.BeadID))
	if err := runCmd(ctx.ProjectDir, "npm", "install", "--offline", "--silent"); err != nil {
		ctx.log("npm-install: offline cache miss, falling back to online")
		if err2 := runCmd(ctx.ProjectDir, "npm", "install", "--silent"); err2 != nil {
			return err2
		}
	}
	return nil
}

func writeComponent(ctx TemplateContext, fm Frontmatter) error {
	ctx.log(fmt.Sprintf("write-component: %s files=%v", ctx.BeadID, fm.Files))
	return writeRegistryFiles(ctx, fm)
}

func buildTpl(ctx TemplateContext, _ Frontmatter) error {
	ctx.log(fmt.Sprintf("build: %s", ctx.BeadID))
	return runCmd(ctx.ProjectDir, "npm", "run", "build")
}

func serveTpl(ctx TemplateContext, _ Frontmatter) error {
	// Preview-server lifecycle is owned by the M2/M3 step beads; the
	// template's job here is just to be a no-op marker.
	ctx.log(fmt.Sprintf("serve: %s (no-op; preview owned by step bead)", ctx.BeadID))
	return nil
}

// errorThenRecoverState — per-bead so concurrent sessions don't race
// over a package-scoped counter. Each bead's first attempt throws;
// subsequent attempts succeed.
var (
	errorThenRecoverMu  sync.Mutex
	errorThenRecoverHit = map[string]bool{}
)

func errorThenRecover(ctx TemplateContext, _ Frontmatter) error {
	errorThenRecoverMu.Lock()
	first := !errorThenRecoverHit[ctx.BeadID]
	errorThenRecoverHit[ctx.BeadID] = true
	errorThenRecoverMu.Unlock()
	ctx.log(fmt.Sprintf("error-then-recover: bead=%s first=%v", ctx.BeadID, first))
	if first {
		return fmt.Errorf("error-then-recover: deliberate first-attempt failure for %s (gm-root.28.5)", ctx.BeadID)
	}
	return nil
}

func noop(ctx TemplateContext, fm Frontmatter) error {
	ctx.log(fmt.Sprintf("noop: %s files=%v", ctx.BeadID, fm.Files))
	return writeRegistryFiles(ctx, fm)
}

// ─── Helpers ──────────────────────────────────────────────────────

func writeRegistryFiles(ctx TemplateContext, fm Frontmatter) error {
	for _, rel := range fm.Files {
		if len(rel) > 0 && rel[len(rel)-1] == '/' {
			// Directory marker (e.g. 'node_modules/'); skip — real
			// commands like npm-install create these.
			continue
		}
		body, ok := resolveContent(rel, ctx.Labels)
		if !ok {
			return fmt.Errorf("mock: no registry entry for file %q (bead %s)", rel, ctx.BeadID)
		}
		if err := writeProjectFile(ctx.ProjectDir, rel, body); err != nil {
			return err
		}
		if rel == "src/App.tsx" && hasLabel(ctx.Labels, "milestone:m3") {
			if err := writeProjectFile(ctx.ProjectDir, "src/App.test.tsx", fileContentAppTestTSXM3); err != nil {
				return err
			}
		}
	}
	return nil
}

func writeProjectFile(projectDir, rel, body string) error {
	abs := filepath.Join(projectDir, rel)
	if err := os.MkdirAll(filepath.Dir(abs), 0o755); err != nil {
		return err
	}
	return os.WriteFile(abs, []byte(body), 0o644)
}

func hasLabel(labels []string, want string) bool {
	for _, label := range labels {
		if label == want {
			return true
		}
	}
	return false
}

func resolveContent(rel string, labels []string) (string, bool) {
	// App.tsx has three versions disambiguated by milestone label.
	if rel == "src/App.tsx" {
		for _, l := range labels {
			if l == "milestone:m3" {
				v, ok := contentRegistry["src/App.tsx#m3"]
				return v, ok
			}
			if l == "milestone:m2" {
				v, ok := contentRegistry["src/App.tsx#m2"]
				return v, ok
			}
		}
		v, ok := contentRegistry["src/App.tsx#m1"]
		return v, ok
	}
	v, ok := contentRegistry[rel]
	return v, ok
}

func runCmd(cwd string, cmd string, args ...string) error {
	c := exec.Command(cmd, args...)
	c.Dir = cwd
	out, err := c.CombinedOutput()
	if err != nil {
		return fmt.Errorf("mock: %s %v failed in %s: %w\n%s", cmd, args, cwd, err, string(out))
	}
	return nil
}

// ─── Content registry ─────────────────────────────────────────────
//
// Hand-tuned to the D17 contract (testids + frontmatter) and the D18
// oracle (numeric correctness, formatted-string °F). If the M3
// oracle fails on a row mismatch, look here first — this is the
// source of truth the mock writes.

var contentRegistry = map[string]string{
	"package.json":                  fileContentPackageJSON,
	"vite.config.ts":                fileContentViteConfig,
	"tsconfig.json":                 fileContentTSConfig,
	"index.html":                    fileContentIndexHTML,
	"src/main.tsx":                  fileContentMainTSX,
	"src/App.tsx#m1":                fileContentAppTSXm1,
	"src/App.tsx#m2":                fileContentAppTSXm2,
	"src/App.tsx#m3":                fileContentAppTSXm3,
	"src/App.test.tsx":              fileContentAppTestTSX,
	"src/temperatureRows.ts":        fileContentTemperatureRows,
	"src/TemperatureTable.tsx":      fileContentTemperatureTable,
	"src/TemperatureTable.test.tsx": fileContentTemperatureTableTest,
}

const fileContentPackageJSON = `{
  "name": "temperature-spa",
  "private": true,
  "version": "0.0.0",
  "type": "module",
  "scripts": {
    "dev": "vite",
    "build": "tsc --noEmit && vite build",
    "preview": "vite preview",
    "test": "vitest run"
  },
  "dependencies": {
    "react": "^18.3.1",
    "react-dom": "^18.3.1"
  },
  "devDependencies": {
    "@testing-library/react": "^16.1.0",
    "@types/react": "^18.3.12",
    "@types/react-dom": "^18.3.1",
    "@vitejs/plugin-react": "^4.3.4",
    "@asamuzakjp/css-color": "^3.2.0",
    "jsdom": "^25.0.1",
    "typescript": "^5.6.3",
    "vite": "^5.4.11",
    "vitest": "^2.1.8"
  }
}
`

const fileContentViteConfig = `import { defineConfig } from 'vite';
import react from '@vitejs/plugin-react';

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [react()],
  test: {
    environment: 'jsdom',
    globals: true,
  },
});
`

const fileContentTSConfig = `{
  "compilerOptions": {
    "target": "ES2022",
    "lib": ["ES2022", "DOM", "DOM.Iterable"],
    "module": "ESNext",
    "moduleResolution": "Bundler",
    "jsx": "react-jsx",
    "strict": true,
    "esModuleInterop": true,
    "resolveJsonModule": true,
    "skipLibCheck": true,
    "isolatedModules": true,
    "noEmit": true
  },
  "include": ["src"]
}
`

const fileContentIndexHTML = `<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Temperature SPA</title>
  </head>
  <body>
    <div id="root"></div>
    <script type="module" src="/src/main.tsx"></script>
  </body>
</html>
`

const fileContentMainTSX = `import React from 'react';
import ReactDOM from 'react-dom/client';
import App from './App';

const root = document.getElementById('root');
if (!root) throw new Error('root element not found');
ReactDOM.createRoot(root).render(<React.StrictMode><App /></React.StrictMode>);
`

const fileContentAppTSXm1 = `export default function App() {
  return <div data-testid="app-root">Scaffold ready</div>;
}
`

const fileContentAppTSXm2 = `export default function App() {
  return <div data-testid="app-root">Hello world</div>;
}
`

const fileContentAppTSXm3 = `import TemperatureTable from './TemperatureTable';

export default function App() {
  return (
    <div data-testid="app-root">
      <TemperatureTable />
    </div>
  );
}
`

const fileContentAppTestTSX = `import { describe, it, expect } from 'vitest';
import { render, screen } from '@testing-library/react';
import App from './App';

describe('App', () => {
  it('renders Hello world inside app-root', () => {
    render(<App />);
    const root = screen.getByTestId('app-root');
    expect(root.textContent).toBe('Hello world');
  });
});
`

const fileContentAppTestTSXM3 = `import { describe, it, expect } from 'vitest';
import { render, screen } from '@testing-library/react';
import App from './App';

describe('App', () => {
  it('keeps app-root and renders the temperature table', () => {
    render(<App />);
    expect(screen.getByTestId('app-root')).toBeTruthy();
    expect(screen.getByTestId('temperature-table')).toBeTruthy();
  });
});
`

const fileContentTemperatureRows = `// 16 rows for °C ∈ {0, 20, 40, ..., 300}. °F = °C × 9/5 + 32, one decimal.
export interface TemperatureRow {
  celsius: number;
  fahrenheit: string;
}

export function temperatureRows(): TemperatureRow[] {
  const rows: TemperatureRow[] = [];
  for (let c = 0; c <= 300; c += 20) {
    rows.push({
      celsius: c,
      fahrenheit: ((c * 9) / 5 + 32).toFixed(1),
    });
  }
  return rows;
}
`

const fileContentTemperatureTable = "import { temperatureRows } from './temperatureRows';\n\nexport default function TemperatureTable() {\n  const rows = temperatureRows();\n  return (\n    <table data-testid=\"temperature-table\">\n      <thead>\n        <tr>\n          <th>Celsius</th>\n          <th>Fahrenheit</th>\n        </tr>\n      </thead>\n      <tbody>\n        {rows.map((r) => (\n          <tr key={r.celsius} data-testid={`row-${r.celsius}`}>\n            <td>{r.celsius}</td>\n            <td>{r.fahrenheit}</td>\n          </tr>\n        ))}\n      </tbody>\n    </table>\n  );\n}\n"

const fileContentTemperatureTableTest = `import { describe, it, expect } from 'vitest';
import { render, screen } from '@testing-library/react';
import TemperatureTable from './TemperatureTable';

describe('TemperatureTable', () => {
  it('renders exactly 16 rows', () => {
    render(<TemperatureTable />);
    const rows = screen.getAllByTestId(/^row-/);
    expect(rows).toHaveLength(16);
  });

  it('row-0 shows 32.0', () => {
    render(<TemperatureTable />);
    const row = screen.getByTestId('row-0');
    expect(row.textContent).toContain('0');
    expect(row.textContent).toContain('32.0');
  });

  it('row-100 shows 212.0', () => {
    render(<TemperatureTable />);
    const row = screen.getByTestId('row-100');
    expect(row.textContent).toContain('100');
    expect(row.textContent).toContain('212.0');
  });

  it('row-300 shows 572.0', () => {
    render(<TemperatureTable />);
    const row = screen.getByTestId('row-300');
    expect(row.textContent).toContain('300');
    expect(row.textContent).toContain('572.0');
  });
});
`
